### DISCUSSION TITLE
[Feature] Juggle point trigger

### DISCUSSION THREAD

**Discussion by Amidweiz** [ikemen-engine/Ikemen-GO#939] [Feature Request] [answered]
**Is your feature request related to a problem? Please describe.**
I've noticed within `ai.zss` within KFMz that in the comments, it notes that there is no way to detect how many juggle points the default juggle system contains. In turn it has to utilize a map value of the assumed count of the juggle points
```
#------------|Juggle point|-----------------------------------------------------
#Currently, juggle points cannot be detected, so this hackish way of doing so will help the AI
#determine what juggles it has left to use for combo attacks
ignoreHitPause if moveHit = 2 && !isHelper && cond(numTarget, target,hitFall, 0) {
	if hitCount = 1 || map(jug) = 0 {
		if stateNo = 200 {
			let jug = 1;
		} else if stateNo = 600 {
			let jug = 2;
		} else if stateNo = 630 {
			let jug = 3;
		} else if stateNo = 230 || stateNo = 210 || stateNo = 610 || stateNo = 640 || stateNo = 1000 || stateNo = 1010 || stateNo = 1050 || stateNo = 1060 || stateNo = 1070 ||
			stateNo = 1100 || stateNo = 1110 || stateNo = 1200 || stateNo = 1210 || stateNo = 1400 || stateNo = 1410 || stateNo = 3000 || stateNo = 3050 {
			let jug = 4;
		} else if stateNo = 240 || stateNo = 400 || stateNo = 430 {
			let jug = 5;
		} else if stateNo = 410 || stateNo = 1020 || stateNo = 1420 || stateNo = 1220 || stateNo = 1120 {
			let jug = 6;
		} else if stateNo = 440 {
			let jug = 7;
		}
		mapAdd{map: "jug"; value: $jug}
	}
}

if enemyNear,moveType != H || enemyNear,stateNo = 5040 || enemyNear,stateNo = [5200, 5210] {
	mapSet{map: "jug"; value: 0}
}
```

**Describe the solution you'd like**
Similar to the `framespercount` trigger, the solution would just be a `jugglepoints` trigger to return said value.

**Comment by Lazin3ss** [comment]
it seems it already exists, though it's not documented in the wiki:

https://github.com/ikemen-engine/Ikemen-GO/blob/b37927dd118b546b27f942ac89d379fb2c8097b3/src/compiler.go#L2545-L2546

https://github.com/ikemen-engine/Ikemen-GO/blob/b37927dd118b546b27f942ac89d379fb2c8097b3/src/bytecode.go#L2012-L2013

**Comment by Amidweiz** [comment]
From testing it 0.98.2, it returns the current amount of juggle points said attack is going to take away from the total rather than the total itself.

**Comment by Amidweiz** [comment]
I think it's current implementation could be renamed `hitjuggle` for the juggle points of a current attack and `jugglepoints` for getting the `airjuggle =` value from `[Data]`

**Comment by Lazin3ss** [comment]
can't that be obtained by using `const(data.airjuggle)`?

**Comment by Amidweiz** [comment]
`const(data.airjuggle)` does work for getting the total airjuggle points. The wiki would need to be updated to account for the missing trigger.

**Comment by Amidweiz** [comment]
I've noticed a bug with the current implementation of the `jugglepoints` trigger. It only can recognize when the `juggle =` parameter is used within the statedef and not the `air.juggle = ` parameter within a hitdef.

**Comment by potsmugen** [comment]
I think a trigger like you're asking for would be better suited as a GetHitVar, since the juggle points are tracked by the character that is being juggled. So the usage would generally be something like "enemynear, gethitvar(juggleorsomething) > X".

Edit: Then again I guess a GetHitVar should be pointing to something that happened in the last hit, so maybe a separate trigger would be better. Like how there's GetHitVar(damage) and also ReceivedDamage. My general message is that remaining juggle points should be read from the victim instead of the attacker.

Edit 2: And come to think of it the way Mugen's juggling works you have *each* attacker tracking the juggle points of *each* target they have individually. So a player may run out of juggle points and have his partner continue the combo. A trigger for juggle points would be more complicated than I thought I guess.

**Comment by Orochikyocr** [comment]
https://github.com/ikemen-engine/Ikemen-GO/issues/382
https://github.com/ikemen-engine/Ikemen-GO/issues/764

The juggle points trigger is not working and it is already reported.

Use the search function here in Github to check if there are other posts about the same feature/issue you want to talk about.

**Comment by K4thos** [comment, marked as answer]
https://github.com/ikemen-engine/Ikemen-GO/wiki/Triggers-(new)#jugglepoints-nightly-build-only
